WordPress/Classes/ViewRelated/Domains/Domain registration/DomainPurchasingWebFlowController.swift

```diff
@@ -50,29 +56,53 @@ final class DomainPurchasingWebFlowController {
     }
 
     func purchase(domain: DomainSuggestion, site: Blog, completion: CompletionHandler? = nil) {
-        let middleware: CompletionHandler = { [weak self] result in
-            if let self = self, case let .failure(error) = result, !error.trusted {
-                let userInfo = self.userInfoForError(error, domain: domain, site: site)
-                self.crashLogger.logError(error, userInfo: userInfo, level: error.level)
+        let middleware = self.middleware(for: completion, domain: domain, site: site)
+        guard let presentingViewController else {
+            middleware(.failure(.internal("The presentingViewController is deallocated")))
+            return
+        }
+        guard let domain = Domain(domain: domain, site: site) else {
+            middleware(.failure(.invalidInput))
+            return
+        }
+        self.createCartAndPresentWebView(domain: domain, in: presentingViewController, completion: middleware)
+    }
+
+    // MARK: - Private: Completion Middleware
+
+    /// Process the domain purchasing result before handing it over to the caller.
+    private func middleware(for completion: CompletionHandler?, domain: DomainSuggestion, site: Blog) -> CompletionHandler {
+        return { [weak self] result in
+            if let self = self {
+                self.handleError(result: result, domain: domain, site: site)
+                self.trackCompletionEvents(result: result, domain: domain, site: site)
             }
             if let completion {
                 DispatchQueue.main.async {
                     completion(result)
                 }
             }
         }
-        guard let presentingViewController else {
-            middleware(.failure(.internal("The presentingViewController is deallocated")))
+    }
+
+    /// Logs the error in case the domain purchasing resulted in a failure.
+    private func handleError(result: Result<String, DomainPurchasingError>, domain: DomainSuggestion, site: Blog) {
+        guard case let .failure(error) = result, !error.trusted else {
             return
         }
-        guard let domain = Domain(domain: domain, site: site) else {
-            middleware(.failure(.invalidInput))
+        let userInfo = self.userInfoForError(error, domain: domain, site: site)
+        self.crashLogger.logError(error, userInfo: userInfo, level: error.level)
+    }
+
+    /// Tracks domain purchasing analytics when it completes.
+    private func trackCompletionEvents(result: Result<String, DomainPurchasingError>, domain: DomainSuggestion, site: Blog) {
+        guard case .success = result else {
             return
         }
-        self.createCartAndPresentWebView(domain: domain, in: presentingViewController, completion: middleware)
+        self.trackDomainPurchasingSucceeded(blog: site)
     }
```

Comment:
The `middleware` and `handleError` methods are just refactors of an existing logic; the `trackCompletionEvents` method is the real new change.

Reply:
Thanks for explaining, and for doing the refactor :+1: